services/core/MasterDriverAgent/tests/test_gpio_driver.py

```diff
@@ -0,0 +1,309 @@
+# completeness, or usefulness or any information, apparatus, product,
+# software, or process disclosed, or represents that its use would not infringe
+# privately owned rights. Reference herein to any specific commercial product,
+# process, or service by trade name, trademark, manufacturer, or otherwise
+# does not necessarily constitute or imply its endorsement, recommendation, or
+# favoring by the United States Government or any agency thereof, or
+# Battelle Memorial Institute. The views and opinions of authors expressed
+# herein do not necessarily state or reflect those of the
+# United States Government or any agency thereof.
+#
+# PACIFIC NORTHWEST NATIONAL LABORATORY operated by
+# BATTELLE for the UNITED STATES DEPARTMENT OF ENERGY
+# under Contract DE-AC05-76RL01830
+# }}}
+
+import time
+import logging
+import threading
+import pytest
+import json
+from  gpiozero.pins.mock import MockFactory
+from services.core.MasterDriverAgent.master_driver.interfaces.gpio import Interface
+
+_log = logging.getLogger(__name__)
+
+# Contents of driver config.
+
+config_dict = {"pin_factory": "mock"}
+
+# Registry CSV Contents:
+#
+# Point Name,Pin,Device Type,Active High,Initial Value,Pull Up,Bounce Time
+# Solenoid,24,output,TRUE,TRUE,,
+# flowSensor,12,digital_input,,,TRUE,
+# flowLED,13,output,TRUE,TRUE,,
+# abortButton,16,digital_input,,,FALSE,
+
+registry_config_str = '[{"Point Name": "Solenoid",' + \
+                        '"Pin": 24,' + \
+                        '"Device Type": "output",' + \
+                        '"Active High": "TRUE",' + \
+                        '"Initial Value": "TRUE"' + \
+                        '},' + \
+                        '{"Point Name": "flowSensor",' + \
+                         '"Pin": 12,' + \
+                         '"Device Type": "digital_input",' + \
+                         '"Pull Up": "TRUE"' + \
+                        '},' + \
+                        '{"Point Name": "flowLED",' + \
+                         '"Pin": 13,' + \
+                         '"Device Type": "digital_output",' + \
+                         '"Active High": "TRUE",' + \
+                        '"Initial Value": "TRUE"' + \
+                        '},' + \
+                        '{"Point Name": "abortButton",' + \
+                         '"Pin": 16,' + \
+                         '"Device Type": "digital_input",' + \
+                         '"Pull Up": "FALSE"' + \
+                        '}]'
+registry_config_str = json.loads(registry_config_str)
+
+
+def drive_low(abort_pin):
+    time.sleep(2)
+    abort_pin.drive_low()
+    time.sleep(2)
+    return time.time()
+
+
+def drive_high(abort_pin):
+    time.sleep(2)
+    abort_pin.drive_high()
+    time.sleep(2)
+    return time.time()
+
+@pytest.fixture()
+def test_configuration():
+    test = Interface()
+    # Test configuration input validation.
+    with pytest.raises(ValueError):
```

Comment:
Lets break out the failure cases into their own test and use pytest.mark.parametrize to improve code-reuse.